### DISCUSSION TITLE
Using FileDialogs for MVVM

### DISCUSSION THREAD

**Discussion by nlogozzo** [AvaloniaUI/Avalonia#5344] [Q&A] [answered]
Hi all,

I'm trying to figure out how to use open file dialogs (and other file dialogs) while keeping the MVVM pattern. Obviously the ViewModel shouldn't know anything about the Views. So i'm creating services to use the file dialogs, however the ShowAsync method of the dialogs requires a Window object, requiring me to pass my Window to the ViewModel and then the Service, which completely breaks the MVVM pattern. Is the anyway I can achieve using dialogs and services without requiring me to pass the Window to the service and not break the pattern?

Thanks in advanced!

**Comment by rstm-sf** [comment]
Hello!

MVVM is not a strict pattern, it just gives us some thought about how best to split the program into component parts :)

Nevertheless, if we want to separate something that should not be in the View and View-Model, then the following approaches exist for this:
- the Behaviors (see [AvaloniaBehaviors](https://github.com/wieslawsoltes/AvaloniaBehaviors))
- the Services via [Command or Method](https://avaloniaui.net/docs/binding/binding-to-commands) ( [Pharmacist](https://www.reactiveui.net/docs/handbook/events/)? )
- the Custom Control

It remains to choose an approach :)

**Reply by nlogozzo** [reply]
Coming from WPF, MVVM is a strict pattern. I do understand how it works and about commands and services, however I'm having the issue of where the Window is required by the FileDialogs thus making it tricky to implement via a service. I might try to pass the window to the service via dependency injection...

**Comment by jp2masa** [comment, marked as answer]
I think the easiest way is to create your own service, which calls Avalonia service, after resolving view from view model, something like this:

```cs
class FileDialogService
{
    public Task<string[]> ShowOpenFileDialogAsync(ViewModelBase parent)
    {
        var window = ResolveViewFromViewModel(parent);
        return new OpenFileDialog(window).ShowAsync(window);
    }
}
```

**Comment by nlogozzo** [comment]
Thanks! This solution worked. Here's how I implemented it.
```csharp
public static class ViewLocator
    {
        /// <summary>
        /// Finds a view from a given ViewModel
        /// </summary>
        /// <param name="vm">The ViewModel representing a View</param>
        /// <returns>The View that matches the ViewModel. Null is no match found</returns>
        public static Window ResolveViewFromViewModel<T>(T vm) where T : ViewModelBase
        {
            var name = vm.GetType().AssemblyQualifiedName!.Replace("ViewModel", "View");
            var type = Type.GetType(name);
            return type != null ? (Window)Activator.CreateInstance(type)! : null;
        }
    }
```

```csharp
/// <summary>
    /// An interface that represents working with IO dialogs
    /// </summary>
    public interface IIODialogService
    {
        /// <summary>
        /// Shows an open file dialog
        /// </summary>
        /// <param name="title">The title for the dialog</param>
        /// <param name="filter">The filter of file types allowed by the dialog</param>
        /// <returns>The path of the file that was opened. Null if no file was opened</returns>
        Task<string> ShowOpenFileDialogAsync(ViewModelBase parent, string title, FileDialogFilter filter);

        /// <summary>
        /// Shows an open folder dialog
        /// </summary>
        /// <param name="title">The title for the dialog</param>
        /// <returns>The path of the folder that was opened. Null if no folder was opened</returns>
        Task<string> ShowOpenFolderDialogAsync(ViewModelBase parent, string title);

        /// <summary>
        /// Shows a save file dialog
        /// </summary>
        /// <param name="title">The title for the dialog</param>
        /// <param name="filter">The filter of file types allowed by the dialog</param>
        /// <returns>The path of the file that was saved. Null if no file was saved</returns>
        Task<string> ShowSaveFileDialogAsync(ViewModelBase parent, string title, FileDialogFilter filter);
    }
```

```csharp
    /// <summary>
    /// A service that contains methods for working with IO dialogs
    /// </summary>
    public class IODialogService : IIODialogService
    {
        /// <summary>
        /// Shows an open file dialog
        /// </summary>
        /// <param name="title">The title for the dialog</param>
        /// <param name="filter">The filter of file types allowed by the dialog</param>
        /// <returns>The path of the file that was opened. Null if no file was opened</returns>
        public async Task<string> ShowOpenFileDialogAsync(ViewModelBase parent, string title, FileDialogFilter filter)
        {
            var openFileDialog = new OpenFileDialog()
            {
                Title = title,
                AllowMultiple = false
            };
            openFileDialog.Filters.Add(filter);
            var result = await openFileDialog.ShowAsync(ViewLocator.ResolveViewFromViewModel(parent));
            return result[0];
        }

        /// <summary>
        /// Shows an open folder dialog
        /// </summary>
        /// <param name="title">The title for the dialog</param>
        /// <returns>The path of the folder that was opened. Null if no folder was opened</returns>
        public async Task<string> ShowOpenFolderDialogAsync(ViewModelBase parent, string title)
        {
            var openFolderDialog = new OpenFolderDialog()
            {
                Title = title
            };
            return await openFolderDialog.ShowAsync(ViewLocator.ResolveViewFromViewModel(parent));
        }

        /// <summary>
        /// Shows a save file dialog
        /// </summary>
        /// <param name="title">The title for the dialog</param>
        /// <param name="filter">The filter of file types allowed by the dialog</param>
        /// <returns>The path of the file that was saved. Null if no file was saved</returns>
        public async Task<string> ShowSaveFileDialogAsync(ViewModelBase parent, string title, FileDialogFilter filter)
        {
            var saveFileDialog = new SaveFileDialog()
            {
                Title = title
            };
            saveFileDialog.Filters.Add(filter);
            return await saveFileDialog.ShowAsync(ViewLocator.ResolveViewFromViewModel(parent));
        }
    }
```

**Reply by blahetal** [reply]
Sorry for bringing the old topic back to life.

How the above code could be right? The method `ResolveViewFromViewModel` returns a new window (created by `(Window)Activator.CreateInstance(type)`), it does not return an instance of an existing window.

**Reply by mysteryx93** [reply]
Correct. It's useful for showing new Windows from your ViewModel. If you want to get existing windows associated with existing ViewModels (such as for getting the Owner of your new window), you can use this

```c#
private static IEnumerable<Window> Windows =>
    (Application.Current?.ApplicationLifetime as IClassicDesktopStyleApplicationLifetime)?.Windows ?? Array.Empty<Window>();

public override Window? FindWindowByViewModel(INotifyPropertyChanged viewModel) =>
    Windows.FirstOrDefault(x => ReferenceEquals(viewModel, x.DataContext));
```

or use the [HanumanInstitute.MvvmDialogs library](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs) that handles this.

**Reply by blahetal** [reply]
Thanks mysteryx93, that was the missing piece of the puzzle but I will use MvvmDialogs anyway. Just got it working, hopefully it will work with avaloniaui 11 when the final version is out :)

**Reply by kekekeks** [reply]
Note that lifetime _is not guaranteed to be available_, especially in embedding scenario. You shouldn't rely on `Application.Current?.ApplicationLifetime` if you are writing library code.

You shouldn't also rely on the lifetime being a classic desktop one, since file dialogs also exist on mobile.

**Reply by mysteryx93** [reply]
Mobile is new; will need to look into that next.

In mobile though, you don't show windows and dialogs, but navigate to pages instead. The ViewModel shouldn't care about those differences though...

Why is `Application.Current?.ApplicationLifetime` not guaranteed to be available? In what cases it won't?

**Comment by grokys** [comment]
Glad you've got it solved! To put another solution out there, ReactiveUI's [Interactions](https://www.reactiveui.net/docs/handbook/interactions/) are often a good way to solve this.

**Reply by FoggyFinder** [reply]
I'd glad if someone provide the sample with `Interactions`. I wasn't able to make it work at all.

**Reply by grokys** [reply]
Here you go - I put together a quick sample: https://github.com/grokys/FileDialogMvvm

Does that help?

**Reply by FoggyFinder** [reply]
Great, thank you. I'll try to find the project where I attempted to use `Interactions` and compare.

**Comment by mysteryx93** [comment]
[MvvmDialogs](https://github.com/FantasticFiasco/mvvm-dialogs) serves that purpose in WPF. [I'm currently porting it to Avalonia.](https://github.com/mysteryx93/mvvm-dialogs)

**Reply by lastlink** [reply]
@mysteryx93 Could you do a sample of the open file dialog? OpenFileDialog : FileDialogBase<OpenFileDialogSettings, string[]>?

**Reply by mysteryx93** [reply]
https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/tree/master/samples/Avalonia/Demo.OpenFileDialog

**Comment by mysteryx93** [comment]
The only interface you need to work with is IDialogService.

Off the top of my head, you'll simply call `dialogService.ShowOpenFileDialogAsync`

The usage is very similar to the original MvvmDialogs so you can refer to that documentation.

**Comment by mysteryx93** [comment]
The library is now officially released: [HanumanInstitute.MvvmDialogs.Avalonia](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs)

**Comment by RespectMathias** [comment]
Right, feel retarded, forgot to set **DataContext** anyway this is my implementation for this solution:
```C#
    private static Window GetMainWindow()
    {
        if (Application.Current!.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            return desktop.MainWindow!;
        }
        return null!;
    }

    private async Task ImportDataAsync()
    {
        var options = new FilePickerOpenOptions
        {
            AllowMultiple = false,
            FileTypeFilter = [
                    new("XML Files")
                    {
                        Patterns = ["*.xml"],
                        AppleUniformTypeIdentifiers = ["public.xml"],
                        MimeTypes = ["application/xml", "text/xml"]
                    }]
        };

        var files = await GetMainWindow().StorageProvider.OpenFilePickerAsync(options);
        var filePaths = files.Select(file => file.TryGetLocalPath()).ToArray();

        if (filePaths == null || filePaths.Length == 0)
        {
            return;
        }

        // Load the source data
        // Replace with whatever loader/thing that requires the file path
        // _marketConditions = _sourceDataManager.LoadSourceData(filePaths[0]!);
    }
```

Felt like i needed to comment since OpenFolderDialog is obsolete.
And because the other solution returned null when I used it.

**Comment by timunie** [comment]
Also worth a look: https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#view-interaction-samples
